### PR TITLE
fix( check if sample is present in statusDB before adding a metrics entry)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Adds functionality to check if a metrics entry has a sample present in statusDB to avoid foreign key constraint errors

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/demultiplex/demux_post_processing.py [hunk 1]
```diff
@@ -162,10 +162,20 @@ def add_sequencing_metrics_to_statusdb(
         self, sample_lane_sequencing_metrics: List[SampleLaneSequencingMetrics]
     ) -> None:
         for metric in sample_lane_sequencing_metrics:
-            if not self.metric_exists_in_status_db(metric):
+            if not self.metric_exists_in_status_db(metric) and self.metric_has_sample_in_statusdb(
```
Comment (seallard) [inline cg/meta/demultiplex/demux_post_processing.py:165]: Nice 👍  We could split this long if statement into several lines to potentially make it more readable, you determine if it is worth:

```python
    for metric in sequencing_metrics:
        metric_exists = self.is_metric_in_database(metric)
        sample_exists = self.is_sample_in_database(metric)
        
        if not metric_exists and sample_exists:
            self.add_metric_to_database(metric)
    
    self.status_db.session.commit()
```

**Review by seallard — APPROVED** [review]
Made some suggestions but it looks good as is 👍

**Comment by ChrOertlin** [comment]
# flow does not fail if sample does not exist but logs a warning.

image

File: cg/meta/demultiplex/demux_post_processing.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/meta/demultiplex/demux_post_processing.py]: Yes it does not do the line skipping, i agree fixed it

File: cg/meta/demultiplex/demux_post_processing.py [hunk 2]
```diff
@@ -162,10 +162,20 @@ def add_sequencing_metrics_to_statusdb(
...
+                metric
+            ):
                 self.add_metric_to_status_db(metric)
         self.status_db.session.commit()
 
+    def metric_has_sample_in_statusdb(self, metric: SampleLaneSequencingMetrics) -> bool:
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/meta/demultiplex/demux_post_processing.py]: indeed there is no need to pass the whole object

**Review by Vince-janv — APPROVED** [review]
nice 👍

**Comment by ChrOertlin** [comment]
stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 40.0.1

prod:
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 40.0.1

**Pull request merged**